src/charm.py

```diff
@@ -37,14 +33,31 @@ def __init__(self, *args):
             self.model.unit.status = ops.BlockedStatus(exc.msg)
             return
         self.smtp = smtp.SmtpProvides(self)
+        self.smtp_legacy = smtp.SmtpProvides(self, relation_name=smtp.LEGACY_RELATION_NAME)
+        self.framework.observe(
+            self.on[smtp.LEGACY_RELATION_NAME].relation_created, self._on_legacy_relation_created
+        )
         self.framework.observe(
-            self.on[LEGACY_RELATION_NAME].relation_created, self._on_relation_created
+            self.on[smtp.DEFAULT_RELATION_NAME].relation_created, self._on_relation_created
         )
         self.framework.observe(self.on.config_changed, self._on_config_changed)
         self.framework.observe(self.on.update_status, self._on_update_status)
 
-    def _on_relation_created(self, _) -> None:
-        """Handle a change to the smtp relation."""
+    def _on_relation_created(self, event: ops.RelationCreatedEvent) -> None:
+        """Handle a change to the smtp relation.
+
+        Args:
+            event: Relation created event.
+        """
+        if secret_id := self._charm_state.password_id:
+            secret = self.model.get_secret(id=secret_id)
```

Comment:
It should be worth catching the exception ops.model.SecretNotFoundError here.

Reply:
Mmm the secret should always be there

Reply:
Well, that would certainly not be expected, so falling into error state seems appropriate to me